controller/eks-cluster-config-handler.go

```diff
@@ -793,27 +792,69 @@ func BuildUpstreamClusterState(name string, clusterState *eks.DescribeClusterOut
 		}
 		ngToAdd := eksv1.NodeGroup{
-			NodegroupName: ng.Nodegroup.NodegroupName,
-			DiskSize:      ng.Nodegroup.DiskSize,
-			Labels:        ng.Nodegroup.Labels,
-			DesiredSize:   ng.Nodegroup.ScalingConfig.DesiredSize,
-			MaxSize:       ng.Nodegroup.ScalingConfig.MaxSize,
-			MinSize:       ng.Nodegroup.ScalingConfig.MinSize,
-			Subnets:       aws.StringValueSlice(ng.Nodegroup.Subnets),
-			Tags:          ng.Nodegroup.Tags,
-			Version:       ng.Nodegroup.Version,
-		}
-		if len(ng.Nodegroup.InstanceTypes) > 0 {
-			ngToAdd.InstanceType = ng.Nodegroup.InstanceTypes[0]
+			NodegroupName:        ng.Nodegroup.NodegroupName,
+			DiskSize:             ng.Nodegroup.DiskSize,
+			Labels:               ng.Nodegroup.Labels,
+			DesiredSize:          ng.Nodegroup.ScalingConfig.DesiredSize,
+			MaxSize:              ng.Nodegroup.ScalingConfig.MaxSize,
+			MinSize:              ng.Nodegroup.ScalingConfig.MinSize,
+			Subnets:              aws.StringValueSlice(ng.Nodegroup.Subnets),
+			Tags:                 ng.Nodegroup.Tags,
+			Version:              ng.Nodegroup.Version,
+			RequestSpotInstances: aws.Bool(aws.StringValue(ng.Nodegroup.CapacityType) == eks.CapacityTypesSpot),
+		}
+
+		if aws.BoolValue(ngToAdd.RequestSpotInstances) {
+			ngToAdd.SpotInstanceTypes = ng.Nodegroup.InstanceTypes
+		}
+
+		if ng.Nodegroup.LaunchTemplate != nil {
+			if managedTemplateID == aws.StringValue(ng.Nodegroup.LaunchTemplate.Id) {
+				// If this is a rancher-managed launch template, then we move the data from the launch template to the node group.
+				launchTemplateRequestOutput, err := ec2Service.DescribeLaunchTemplateVersions(&ec2.DescribeLaunchTemplateVersionsInput{
+					LaunchTemplateId: ng.Nodegroup.LaunchTemplate.Id,
+					Versions:         []*string{ng.Nodegroup.LaunchTemplate.Version},
+				})
+				if err != nil || len(launchTemplateRequestOutput.LaunchTemplateVersions) == 0 {
+					return nil, "", fmt.Errorf("error getting launch template info for node group [%s] in cluster [%s]", aws.StringValue(ngToAdd.NodegroupName), upstreamSpec.DisplayName)
+				}
+				launchTemplateData := launchTemplateRequestOutput.LaunchTemplateVersions[0].LaunchTemplateData
+
+				ngToAdd.DiskSize = launchTemplateData.BlockDeviceMappings[0].Ebs.VolumeSize
+				ngToAdd.Ec2SshKey = launchTemplateData.KeyName
+				ngToAdd.ImageID = launchTemplateData.ImageId
+				ngToAdd.InstanceType = launchTemplateData.InstanceType
+				ngToAdd.UserData = launchTemplateData.UserData
+				ngToAdd.ResourceTags = utils.GetInstanceTags(launchTemplateData.TagSpecifications)
+
+				if includeManagedLaunchTemplate {
+					var version *int64
+					versionNumber, err := strconv.ParseInt(aws.StringValue(ng.Nodegroup.LaunchTemplate.Version), 10, 64)
+					if err == nil {
+						version = aws.Int64(versionNumber)
+					}
+
+					ngToAdd.LaunchTemplate = &eksv1.LaunchTemplate{
+						ID:      ng.Nodegroup.LaunchTemplate.Id,
+						Name:    ng.Nodegroup.LaunchTemplate.Name,
+						Version: version,
+					}
```

Comment:
I think if an error occurs we have an uninitialized version.

Reply:
Then EKS will just use the default version, which is the desired outcome.

Reply:
What happens if `if err == nil` is false? Isn't `version` uninitialized at that point, and then `Version: version` will be unset?

Reply:
Yes, and EKS will use the default version of the template, which is the desired result.